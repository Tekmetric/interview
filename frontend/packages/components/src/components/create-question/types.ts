export interface CreateQuestionFormFields {
  title: string
  description: string
}
