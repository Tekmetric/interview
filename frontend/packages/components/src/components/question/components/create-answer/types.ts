export interface CreateAnswerFormFields {
  description: string
}
