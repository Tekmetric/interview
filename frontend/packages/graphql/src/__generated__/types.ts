export type Maybe<T> = T | null;
export type InputMaybe<T> = Maybe<T>;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type MakeEmpty<T extends { [key: string]: unknown }, K extends keyof T> = { [_ in K]?: never };
export type Incremental<T> = T | { [P in keyof T]?: P extends ' $fragmentName' | '__typename' ? T[P] : never };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: { input: string; output: string; }
  String: { input: string; output: string; }
  Boolean: { input: boolean; output: boolean; }
  Int: { input: number; output: number; }
  Float: { input: number; output: number; }
  DateTime: { input: any; output: any; }
};

export type Answer = {
  __typename?: 'Answer';
  author: Author;
  authorId: Scalars['String']['output'];
  createdAt: Scalars['DateTime']['output'];
  description: Scalars['String']['output'];
  id: Scalars['String']['output'];
};

export type Author = {
  __typename?: 'Author';
  firstName: Scalars['String']['output'];
  id: Scalars['String']['output'];
  lastName: Scalars['String']['output'];
};

export type CreateAnswerDto = {
  description: Scalars['String']['input'];
  questionId: Scalars['String']['input'];
};

export type CreateQuestionDto = {
  description: Scalars['String']['input'];
  title: Scalars['String']['input'];
};

export type Mutation = {
  __typename?: 'Mutation';
  createAnswer: Answer;
  createQuestion: Question;
  login?: Maybe<Profile>;
  logout?: Maybe<Scalars['String']['output']>;
  resolveQuestion: Question;
};


export type MutationCreateAnswerArgs = {
  input: CreateAnswerDto;
};


export type MutationCreateQuestionArgs = {
  input: CreateQuestionDto;
};


export type MutationLoginArgs = {
  email: Scalars['String']['input'];
  password: Scalars['String']['input'];
};


export type MutationResolveQuestionArgs = {
  id: Scalars['String']['input'];
};

export type Profile = {
  __typename?: 'Profile';
  fullName: Scalars['String']['output'];
  role: UserRole;
  userId: Scalars['String']['output'];
};

export type Query = {
  __typename?: 'Query';
  hello: Scalars['String']['output'];
  profile: Profile;
  question: Question;
  questions: Array<Question>;
};


export type QueryQuestionArgs = {
  id: Scalars['String']['input'];
};


export type QueryQuestionsArgs = {
  status: QuestionStatus;
};

export type Question = {
  __typename?: 'Question';
  answers: Array<Answer>;
  author: Author;
  authorId: Scalars['String']['output'];
  createdAt: Scalars['DateTime']['output'];
  description: Scalars['String']['output'];
  id: Scalars['String']['output'];
  permissions: QuestionPermissions;
  status: QuestionStatus;
  title: Scalars['String']['output'];
};

export type QuestionPermissions = {
  __typename?: 'QuestionPermissions';
  canResolve: Scalars['Boolean']['output'];
  id: Scalars['String']['output'];
};

export enum QuestionStatus {
  Completed = 'COMPLETED',
  Pending = 'PENDING'
}

export enum UserRole {
  Admin = 'ADMIN',
  User = 'USER'
}

export type LoginMutationVariables = Exact<{
  email: Scalars['String']['input'];
  password: Scalars['String']['input'];
}>;


export type LoginMutation = { __typename?: 'Mutation', login?: { __typename?: 'Profile', userId: string } | null };
