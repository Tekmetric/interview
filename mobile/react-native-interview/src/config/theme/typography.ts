const typography = {
  fontSize: {
    xs: 12,
    sm: 14,
    md: 16,
    lg: 20,
    xl: 24,
    xxl: 32,
  },
  fontWeight: {
    regular: 400,
    medium: 500,
    bold: 700,
  },
};

export default typography;
