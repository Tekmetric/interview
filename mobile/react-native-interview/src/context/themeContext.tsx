import { DarkTheme, DefaultTheme, Theme } from '@react-navigation/native';
import { createContext, useState, useContext, useMemo } from 'react';
import { type ColorSchemeName, useColorScheme } from 'react-native';

type ThemeContextType = {
  theme: Theme;
  mode: ColorSchemeName;
  toggleMode: () => void;
};

const ThemeContext = createContext<ThemeContextType | null>(null);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const [mode, setMode] = useState<ColorSchemeName>(useColorScheme() ?? 'light');

  const theme = useMemo(() => {
    return mode === 'dark' ? DarkTheme : DefaultTheme;
  }, [mode]);

  const toggleMode = () => {
    setMode((prevMode) => (prevMode === 'light' ? 'dark' : 'light'));
  };

  const value = useMemo(
    () => ({
      theme,
      mode,
      toggleMode,
    }),
    [theme, mode],
  );

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
};

export const useTheme = (): ThemeContextType => {
  const context = useContext(ThemeContext);

  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};
