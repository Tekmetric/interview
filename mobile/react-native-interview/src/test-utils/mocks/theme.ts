import { Theme } from '@react-navigation/native';

export const mockTheme: Theme = {
  dark: false,
  colors: {
    primary: '#000000',
    background: '#FFFFFF',
    card: '#FFFFFF',
    text: '#000000',
    border: '#000000',
    notification: '#000000',
  },
  fonts: {
    regular: {
      fontFamily: 'System',
      fontWeight: '400' as const,
    },
    medium: {
      fontFamily: 'System',
      fontWeight: '500' as const,
    },
    bold: {
      fontFamily: 'System',
      fontWeight: '700' as const,
    },
    heavy: {
      fontFamily: 'System',
      fontWeight: '800' as const,
    },
  },
};
