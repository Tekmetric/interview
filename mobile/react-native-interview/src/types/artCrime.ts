export type ArtCrimeFilters = {
  crimeCategory?: string;
  maker?: string;
  materials?: string;
  period?: string;
  idInAgency?: string;
  referenceNumber?: string;
  measurements?: string;
  additionalData?: string;
};

export type ArtCrimeSorting = {
  sort_order?: 'desc' | 'asc'; // defaults to 'desc'
  // Note: at the moment 'publication' will return [], it defaults to 'modified'..and '_score' i'm not sure how it sorts
  sort_on?: 'modified' | 'publication' | '_score';
};

export type ArtCrimeQueryParams = {
  page?: number;
  pageSize?: number;
  title?: string;
} & ArtCrimeFilters &
  ArtCrimeSorting;
