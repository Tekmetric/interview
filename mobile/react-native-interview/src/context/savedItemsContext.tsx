import React, { createContext, useContext, useState, useCallback, useEffect, useMemo } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ArtCrime } from '@/types/artCrime';
const SAVED_ITEMS_KEY = '@saved_items';

type SavedItemsContextType = {
  savedItems: ArtCrime[];
  isLoading: boolean;
  saveItem: (item: ArtCrime) => Promise<void>;
  removeItem: (itemId: string) => Promise<void>;
  isItemSaved: (itemId: string) => boolean;
};

const SavedItemsContext = createContext<SavedItemsContextType | null>(null);

export function SavedItemsProvider({ children }: { children: React.ReactNode }) {
  const [savedItems, setSavedItems] = useState<ArtCrime[]>([]);
  const [isLoading, setIsLoading] = useState(true);

  const loadSavedItems = useCallback(async () => {
    try {
      const savedItemsJson = await AsyncStorage.getItem(SAVED_ITEMS_KEY);
      if (savedItemsJson) {
        setSavedItems(JSON.parse(savedItemsJson));
      } else {
        setSavedItems([]);
      }
    } catch (error) {
      console.error('Error loading saved items:', error);
    } finally {
      setIsLoading(false);
    }
  }, []);

  useEffect(() => {
    loadSavedItems();
  }, [loadSavedItems]);

  const saveItem = useCallback(
    async (item: ArtCrime) => {
      try {
        const updatedItems = [...savedItems, item];
        await AsyncStorage.setItem(SAVED_ITEMS_KEY, JSON.stringify(updatedItems));
        setSavedItems(updatedItems);
      } catch (error) {
        console.error('Error saving item:', error);
      }
    },
    [savedItems],
  );

  const removeItem = useCallback(
    async (itemId: string) => {
      try {
        const updatedItems = savedItems.filter((item) => item.uid !== itemId);
        await AsyncStorage.setItem(SAVED_ITEMS_KEY, JSON.stringify(updatedItems));
        setSavedItems(updatedItems);
      } catch (error) {
        console.error('Error removing item:', error);
      }
    },
    [savedItems],
  );

  const isItemSaved = useCallback(
    (itemId: string) => {
      return savedItems.some((item) => item.uid === itemId);
    },
    [savedItems],
  );

  const value = useMemo(
    () => ({
      savedItems,
      isLoading,
      saveItem,
      removeItem,
      isItemSaved,
    }),
    [savedItems, isLoading, saveItem, removeItem, isItemSaved],
  );

  return <SavedItemsContext.Provider value={value}>{children}</SavedItemsContext.Provider>;
}

export function useSavedItems() {
  const context = useContext(SavedItemsContext);
  if (!context) {
    throw new Error('useSavedItems must be used within a SavedItemsProvider');
  }
  return context;
}
