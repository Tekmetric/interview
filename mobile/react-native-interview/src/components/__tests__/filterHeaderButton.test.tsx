import React from 'react';
import { render } from '@testing-library/react-native';
import FilterHeaderButton from '../buttons/filterHeaderButton';
import { ThemeProvider as NavigationThemeProvider } from '@react-navigation/native';
import { ArtCrimeFilters } from '../../types/artCrime';
import { mockTheme } from '../../test-utils/mocks/theme';
import { ThemeProvider } from '@/context/themeContext';

describe('FilterHeaderButton', () => {
  it('renders correctly with no filters', () => {
    const mockFilters: ArtCrimeFilters = {};

    const { toJSON } = render(
      <ThemeProvider>
        <NavigationThemeProvider value={mockTheme}>
          <FilterHeaderButton filters={mockFilters} onPress={() => {}} />
        </NavigationThemeProvider>
      </ThemeProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with active filters', () => {
    const mockFilters: ArtCrimeFilters = {
      crimeCategory: 'theft',
      maker: 'Picasso',
      period: 'Modern',
    };

    const { toJSON } = render(
      <ThemeProvider>
        <NavigationThemeProvider value={mockTheme}>
          <FilterHeaderButton filters={mockFilters} onPress={() => {}} />
        </NavigationThemeProvider>
      </ThemeProvider>,
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
