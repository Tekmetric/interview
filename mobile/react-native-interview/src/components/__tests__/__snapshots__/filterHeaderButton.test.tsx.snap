// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterHeaderButton renders correctly with active filters 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "marginHorizontal": 16,
      },
      undefined,
    ]
  }
  testID="header-button"
>
  <Feather
    color="rgb(28, 28, 30)"
    name="filter"
    size={24}
  />
  <View
    style={
      {
        "backgroundColor": "red",
        "borderRadius": 4,
        "height": 8,
        "position": "absolute",
        "right": -4,
        "top": -4,
        "width": 8,
      }
    }
  />
</View>
`;

exports[`FilterHeaderButton renders correctly with no filters 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "marginHorizontal": 16,
      },
      undefined,
    ]
  }
  testID="header-button"
>
  <Feather
    color="rgb(28, 28, 30)"
    name="filter"
    size={24}
  />
</View>
`;
